### PR TITLE
Feature/elastic error handling

### PR DESCRIPTION
**Pull request by olovy**
New dependency: resilience4j (https://resilience4j.readme.io/)
New dependency: nebula.lint for checking gradle dependecies (https://github.com/nebula-plugins/gradle-lint-plugin)

Removed unused dependencies and made undeclared uses of transitive dependencies explicit.

Example output from metrics:
```
# HELP api_requests_total Total requests to API.
# TYPE api_requests_total counter
api_requests_total{method="GET",} 152.0
# HELP resilience4j_retry_calls The number of calls
# TYPE resilience4j_retry_calls gauge
resilience4j_retry_calls{name="ElasticClient",kind="successful_without_retry",} 142.0
resilience4j_retry_calls{name="ElasticClient",kind="successful_with_retry",} 0.0
resilience4j_retry_calls{name="ElasticClient",kind="failed_without_retry",} 0.0
resilience4j_retry_calls{name="ElasticClient",kind="failed_with_retry",} 0.0
resilience4j_retry_calls{name="http://localhost:9200",kind="successful_without_retry",} 142.0
resilience4j_retry_calls{name="http://localhost:9200",kind="successful_with_retry",} 0.0
resilience4j_retry_calls{name="http://localhost:9200",kind="failed_without_retry",} 0.0
resilience4j_retry_calls{name="http://localhost:9200",kind="failed_with_retry",} 0.0
# HELP resilience4j_circuitbreaker_calls Total number of calls by kind
# TYPE resilience4j_circuitbreaker_calls histogram
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="0.005",} 121.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="0.01",} 133.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="0.025",} 136.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="0.05",} 139.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="0.075",} 141.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="0.1",} 141.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="0.25",} 142.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="0.5",} 142.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="0.75",} 142.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="1.0",} 142.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="2.5",} 142.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="5.0",} 142.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="7.5",} 142.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="10.0",} 142.0
resilience4j_circuitbreaker_calls_bucket{name="http://localhost:9200",kind="successful",le="+Inf",} 142.0
resilience4j_circuitbreaker_calls_count{name="http://localhost:9200",kind="successful",} 142.0
resilience4j_circuitbreaker_calls_sum{name="http://localhost:9200",kind="successful",} 0.8602747889999998
# HELP resilience4j_circuitbreaker_state The state of the circuit breaker:
# TYPE resilience4j_circuitbreaker_state gauge
resilience4j_circuitbreaker_state{name="http://localhost:9200",state="disabled",} 0.0
resilience4j_circuitbreaker_state{name="http://localhost:9200",state="closed",} 1.0
resilience4j_circuitbreaker_state{name="http://localhost:9200",state="open",} 0.0
resilience4j_circuitbreaker_state{name="http://localhost:9200",state="forced_open",} 0.0
resilience4j_circuitbreaker_state{name="http://localhost:9200",state="half_open",} 0.0
# HELP resilience4j_circuitbreaker_buffered_calls The number of buffered calls
# TYPE resilience4j_circuitbreaker_buffered_calls gauge
resilience4j_circuitbreaker_buffered_calls{name="http://localhost:9200",kind="successful",} 100.0
resilience4j_circuitbreaker_buffered_calls{name="http://localhost:9200",kind="failed",} 0.0
# HELP resilience4j_circuitbreaker_slow_calls The number of slow calls
# TYPE resilience4j_circuitbreaker_slow_calls gauge
resilience4j_circuitbreaker_slow_calls{name="http://localhost:9200",kind="successful",} 0.0
resilience4j_circuitbreaker_slow_calls{name="http://localhost:9200",kind="failed",} 0.0
```

### CODE REVIEW

**Comment by olovy** [comment]
> I feel that this is a fairly large and complicated/complicating change for not-that-big a benefit. But that's just an oppinion, not a fact.
> 
> The dependency changes are great regardless!

Yes I guess the complication part is subjective. But it will make 100% of requests succeed as long as any elastic node is responding.

File: whelk-core/src/main/groovy/whelk/component/ElasticClient.groovy [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+import org.apache.http.impl.client.DefaultHttpClient
+import org.apache.http.impl.conn.PoolingClientConnectionManager
+import org.apache.http.params.HttpConnectionParams
+import org.apache.http.params.HttpParams
+import org.apache.http.util.EntityUtils
+import whelk.exception.WhelkRuntimeException
+
+@Log
+@CompileStatic
+class ElasticClient {
```
Comment (lrosenstrom) [inline whelk-core/src/main/groovy/whelk/component/ElasticClient.groovy]: :+1:

File: whelk-core/src/main/groovy/whelk/component/ElasticClient.groovy [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+                throw new RuntimeException(e.getMessage(), e)
+            }
+            finally {
+                request.reset()
+                request.releaseConnection()
+            }
+        }
+
+        private Tuple2<Integer, String> sendRequestRetry429(HttpRequestBase request) {
+            int backOffTime = 1
```
Comment (lrosenstrom) [inline whelk-core/src/main/groovy/whelk/component/ElasticClient.groovy]: Add unit to variable name

Comment (olovy) [reply whelk-core/src/main/groovy/whelk/component/ElasticClient.groovy]: :+1:

Comment (olovy) [reply whelk-core/src/main/groovy/whelk/component/ElasticClient.groovy]: Could use reslience4j backoff mechanism here too. Right now it's used moved from the old location.

**Pull request merged**